src/match.c

```diff
@@ -1747,23 +1823,74 @@ r_obj* expand_compact_indices(const int* v_o_haystack,
       }
     }
 
-    if (check_for_multiple) {
+    if (check_multiple_needles) {
       if (loc < size_needles) {
-        any_multiple = size_match > 1;
+        any_multiple_needles = size_match > 1;
       } else {
         // Guaranteed second match if in the "extra" matches section
-        any_multiple = true;
+        any_multiple_needles = true;
       }
 
-      if (any_multiple) {
-        if (multiple == VCTRS_MULTIPLE_error) {
-          stop_matches_multiple(loc_needles, needles_arg, haystack_arg, error_call);
-        } else if (multiple == VCTRS_MULTIPLE_warning) {
-          warn_matches_multiple(loc_needles, needles_arg, haystack_arg, error_call);
+      if (any_multiple_needles) {
+        loc_first_multiple_needles = loc_needles;
+
+        switch (relationship) {
+        case VCTRS_RELATIONSHIP_one_to_one:
+          stop_matches_relationship_one_to_one(
+            loc_first_multiple_needles,
+            "needles",
+            needles_arg,
+            haystack_arg,
+            error_call
+          );
+        case VCTRS_RELATIONSHIP_many_to_one:
+          stop_matches_relationship_many_to_one(
+            loc_first_multiple_needles,
+            needles_arg,
+            haystack_arg,
+            error_call
+          );
+        case VCTRS_RELATIONSHIP_warn_many_to_many: {
+          if (any_multiple_haystack) {
+            warn_matches_relationship_many_to_many(
+              loc_first_multiple_needles,
+              loc_first_multiple_haystack,
+              needles_arg,
+              haystack_arg,
+              error_call
+            );
+          }
+          break;
+        }
+        default: {
+          switch (multiple) {
+          case VCTRS_MULTIPLE_all:
+            break;
+          // TODO: Remove deprecated support for `multiple = "error"/"warning"`
+          case VCTRS_MULTIPLE_error:
+            stop_matches_multiple(
+              loc_first_multiple_needles,
+              needles_arg,
+              haystack_arg,
+              error_call
+            );
+          case VCTRS_MULTIPLE_warning: {
+            warn_matches_multiple(
+              loc_first_multiple_needles,
+              needles_arg,
+              haystack_arg,
+              error_call
+            );
+            break;
```

Comment:
Squeezing in some backwards compatible support for `multiple = "error" / "warning"` here in the "we detected multiple matches in `needles`" case